lib/devise/models/confirmable.rb

```diff
@@ -169,6 +169,7 @@ def skip_reconfirmation!
         # in models to map to a nice sign up e-mail.
         def send_on_create_confirmation_instructions
           send_confirmation_instructions
+          skip_reconfirmation!
```

Comment:
@printercu @ulissesalmeida  I don't think it's an app-specific case since there had been a test for it https://github.com/lmduc/devise/blob/7d10019d2a0feaba26af875ed7a7b3e3f4a1a05d/test/models/confirmable_test.rb#L503 . I tried to improve the code of it and mistakenly created a bad behavior, which caused your tests failed. In reality, there aren't many chances for users to face the same problem because there aren't many situations that the system will update emails of users right after creating their accounts. If a user updates his email, the system will find the user from the database, which will set `@bypass_confirmation_postpone` to be `false` and it will works properly. In your test, you update the email of the user right after creating him, then `@bypass_confirmation_postpone` is `true`, which cause the fail. But if you load the user from the database (which happens in real life), `@bypass_confirmation_postpone` is `false` and it works properly. 

Regarding the behavior of updating email from `nil` to a new email, I agree that the new email should be updated to `email` instead of `unconfirmed_email` since it's familiar with creating a new user.